passes/hierarchy/hierarchy.cc

```diff
@@ -710,6 +711,411 @@ RTLIL::Wire *find_implicit_port_wire(Module *module, Cell *cell, const std::stri
+//
+// If mod is not null, it is the current module. Otherwise, this binding was at
+// top-level in the design. counts is a dictionary mapping cell types to the
+// number of places that a cell is instantiated in other modules.
+//
+// If this binding applies to a module (rather than a specific cell
+// instantiating that module), we can modify the module, operating in place. If
+// the binding applies to just one instance, we might have to make a copy
+// (because a version of the module without something bound in might be
+// instantiated somewhere else). We can tell whether that's needed by looking
+// at the counts dictionary. If the count is at most 1, we can still operate in
+// place: the module either only appears at top-level or only appears as the
+// instance into which we're binding. If the count is 2 or greater, we need to
+// take a copy. In this case, we take the copy and modify it. We then decrement
+// the count for the original and give the new copy a count of 1.
+//
+// Finally, this function updates top_mod if *top_mod is non-null and **top_mod
+// is a module that was updated.
+static bool
+insert_binding(RTLIL::Design               &design,
+               dict<RTLIL::IdString, int>  &counts,
+               RTLIL::Module               *mod,
+               const RTLIL::Binding        &binding,
+               RTLIL::Module              **top_mod)
+{
+	std::string desc = binding.describe();
+
+	// If we're not at top-level, look for a cell in mod with the right
+	// (hierarchical) name. This is what section 23.3.1 of IEEE 1800 refers
+	// to as "the local A.B.C".
+	if (mod) {
+		RTLIL::Binding::Path path = binding.find_rel_cell(design, *mod);
+		if (path.size()) {
+			log("Applying %s at a local path.\n", desc.c_str());
+			return bind_path(design, counts, path, binding, top_mod);
+		}
+	}
+
+	// If no luck, try to interpret the name as a hierarchical reference
+	// starting at the top level.
+	std::vector<RTLIL::Binding::Path> paths = binding.find_tl_cells(design);
+	if (!paths.empty()) {
+		bool did_something = false;
+		log("Applying %s at %d top-level path%s.\n",
+		    desc.c_str(), GetSize(paths), paths.size() > 1 ? "s" : "");
+		for (const auto &path : paths)
+			did_something |=
+				bind_path(design, counts, path, binding, top_mod);
+		return did_something;
+	}
+
+	// If neither of these worked, search for a module by name. Note that
+	// this takes no account of design hierarchy.
+	std::vector<RTLIL::Module *> mods = binding.find_concrete_module_targets(design);
+	if (!mods.empty()) {
+		// We can add the binding directly to the module, operating in place.
+		bool did_something = false;
+		log("Applying %s to %d module%s.\n",
+		    desc.c_str(), GetSize(mods), mods.size() > 1 ? "s" : "");
+		for (auto mod : mods)
+			did_something |= binding.bind_into(design, *mod);
+		return did_something;
+	}
+
+	log_error("Couldn't find a target for %s.\n", desc.c_str());
```

Comment:
Does the bind directive have a source location we could log with this error? If so, please update this and the other `log_error`s added elsewhere.

Reply:
It might be nice to add some `logger -expect error` test cases for these errors.

Reply:
Oh, good idea. I'll do that.